tiledb/sm/filesystem/vfs.cc

```diff
@@ -481,10 +481,33 @@ void VFS::copy_dir(const URI& old_uri, const URI& new_uri) {
   auto instrument = make_log_duration_instrument(old_uri, new_uri);
   auto& old_fs = get_fs(old_uri);
   auto& new_fs = get_fs(new_uri);
+  auto src_parent_path = old_uri.to_string();
+  auto dst_parent_path = new_uri.to_string();
+
   if (&old_fs == &new_fs) {
     old_fs.copy_dir(old_uri, new_uri);
   } else {
-    throw UnsupportedOperation("copy_dir");
+    // Create the new directory on the destination FS.
+    new_fs.create_dir(new_uri);
+
+    // Recursively list and copy all source files
+    ResultFilterV2 result_filter = [](const std::string_view&, uint64_t, bool) {
+      return true;  // don't filter out any results.
+    };
+    auto paths = old_fs.ls_filtered_v2(old_uri, result_filter, true);
+
+    for (auto& path : paths) {
+      auto old_path = URI(path.first);
+      auto new_path =
+          URI(dst_parent_path + path.first.substr(src_parent_path.size()));
+      if (old_fs.is_dir(old_path)) {
+        // Create new directories on the destination FS.
+        new_fs.create_dir(new_path);
```

Comment:
It might be a better idea to filter out directories and have `std::filesystem::copy_file` create any missing parent directories.